apollo-runtime-kotlin/src/commonMain/kotlin/com/apollographql/apollo/interceptor/OauthInterceptor.kt

```diff
@@ -0,0 +1,63 @@
+package com.apollographql.apollo.interceptor
+
+import com.apollographql.apollo.ApolloError
+import com.apollographql.apollo.ApolloException
+import com.apollographql.apollo.api.ApolloExperimental
+import com.apollographql.apollo.api.Response
+import com.apollographql.apollo.network.HttpExecutionContext
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.single
+
+@ApolloExperimental
+class OauthInterceptor(private val accessTokenProvider: AccessTokenProvider) : ApolloRequestInterceptor {
+  private fun <T> ApolloRequest<T>.withHeader(name: String, value: String): ApolloRequest<T> {
+    val httpRequestContext = executionContext[HttpExecutionContext.Request]
+        ?: HttpExecutionContext.Request(emptyMap())
+            .let {
+              it.copy(headers = it.headers + (name to value))
+            }
+
+    return ApolloRequest(
+        operation = operation,
+        scalarTypeAdapters = scalarTypeAdapters,
+        executionContext = executionContext + httpRequestContext
+    )
+  }
+
+  private suspend fun <T> proceedWithToken(request: ApolloRequest<T>, interceptorChain: ApolloInterceptorChain, token: String): Response<T> {
+    val newRequest = request.withHeader("Authorization", "Bearer $token")
```

Comment:
Also I think we should rename this interceptor to explicitly provide that this is Oauth with bearer tokenization.
Right now it can give impression that we handle any sort of Oauth authorization schema.

Reply:
Agreed, I renamed that to `BearerTokenInterceptor`